src/encoding.rs

```diff
@@ -0,0 +1,197 @@
+    #[inline]
+    pub fn decode<'b>(&self, bytes: &'b [u8]) -> Result<Cow<'b, str>> {
+        Ok(Cow::Borrowed(std::str::from_utf8(bytes)?))
+    }
+
+    /// Decodes a slice regardless of XML declaration with BOM removal if
+    /// it is present in the `bytes`.
+    ///
+    /// Returns an error in case of malformed sequences in the `bytes`.
+    ///
+    /// If you instead want to use XML declared encoding, use the `encoding` feature
+    pub fn decode_with_bom_removal<'b>(&self, bytes: &'b [u8]) -> Result<Cow<'b, str>> {
+        let bytes = if bytes.starts_with(&[0xEF, 0xBB, 0xBF]) {
+            &bytes[3..]
+        } else {
+            bytes
+        };
+        self.decode(bytes)
+    }
+}
+
+#[cfg(feature = "encoding")]
+impl Decoder {
+    /// Returns the `Reader`s encoding.
+    ///
+    /// This encoding will be used by [`decode`].
+    ///
+    /// [`decode`]: Self::decode
+    pub fn encoding(&self) -> &'static Encoding {
+        self.encoding
+    }
+
+    /// Decodes specified bytes using encoding, declared in the XML, if it was
+    /// declared there, or UTF-8 otherwise, and ignoring BOM if it is present
+    /// in the `bytes`.
+    ///
+    /// Returns an error in case of malformed sequences in the `bytes`.
+    pub fn decode<'b>(&self, bytes: &'b [u8]) -> Result<Cow<'b, str>> {
+        decode(bytes, self.encoding)
+    }
+
+    /// Decodes a slice with BOM removal if it is present in the `bytes` using
+    /// the reader encoding.
+    ///
+    /// If this method called after reading XML declaration with the `"encoding"`
+    /// key, then this encoding is used, otherwise UTF-8 is used.
+    ///
+    /// If XML declaration is absent in the XML, UTF-8 is used.
+    ///
+    /// Returns an error in case of malformed sequences in the `bytes`.
+    pub fn decode_with_bom_removal<'b>(&self, bytes: &'b [u8]) -> Result<Cow<'b, str>> {
+        self.decode(remove_bom(bytes, self.encoding))
+    }
+}
+
+/// Decodes the provided bytes using the specified encoding, ignoring the BOM
+/// if it is present in the `bytes`.
+///
+/// Returns an error in case of malformed sequences in the `bytes`.
+#[cfg(feature = "encoding")]
+pub fn decode<'b>(bytes: &'b [u8], encoding: &'static Encoding) -> Result<Cow<'b, str>> {
+    encoding
+        .decode_without_bom_handling_and_without_replacement(bytes)
+        .ok_or(Error::NonDecodable(None))
+}
+
+/// Decodes a slice with an unknown encoding, removing the BOM if it is present
+/// in the bytes.
+///
+/// Returns an error in case of malformed sequences in the `bytes`.
+#[cfg(feature = "encoding")]
+pub fn decode_with_bom_removal<'b>(bytes: &'b [u8]) -> Result<Cow<'b, str>> {
+    if let Some(encoding) = detect_encoding(bytes) {
+        let bytes = remove_bom(bytes, encoding);
+        decode(bytes, encoding)
+    } else {
+        decode(bytes, UTF_8)
+    }
+}
+
+#[cfg(feature = "encoding")]
+fn split_at_bom<'b>(bytes: &'b [u8], encoding: &'static Encoding) -> (&'b [u8], &'b [u8]) {
```

Comment:
This seems redundant, because first part is not used anywhere. Why just cut off the beginning, as was before, is not enough?

Reply:
You're right, but I was thinking we may want to add some variants that return the BOM the same way that we provide `StartText`.  Or at the very least that it would be useful for testing.

On the other hand, I kinda feel like both `StartText` and returning the BOM have limited utility in practice.  But it feels like an open question.

I'll leave it as-is for now but I wouldn't be be upset if we end up removing it later.